internal/project/watch.go

```diff
@@ -60,3 +65,267 @@ func (w *watchedFiles[T]) update(ctx context.Context, newData T) (updated bool,
 	w.watcherID = watcherID
 	return true, nil
 }
+
+func createGlobMapper(host ProjectHost) func(data map[tspath.Path]string) []string {
+	rootDir := host.GetCurrentDirectory()
+	rootPath := tspath.ToPath(rootDir, "", host.FS().UseCaseSensitiveFileNames())
+	rootPathComponents := tspath.GetPathComponents(string(rootPath), "")
+	isRootWatchable := canWatchDirectoryOrFile(rootPathComponents)
+
+	return func(data map[tspath.Path]string) []string {
+		start := time.Now()
+
+		type glob struct {
+			dir          string
+			nonRecursive bool
+		}
+
+		var globSet core.Set[glob]
+
+		for path, fileName := range data {
+			w := getDirectoryToWatchFailedLookupLocation(
+				fileName,
+				path,
+				rootDir,
+				rootPath,
+				rootPathComponents,
+				isRootWatchable,
+				rootDir,
+				true,
+			)
+			if w == nil {
+				continue
+			}
+			globSet.Add(glob{dir: w.dir, nonRecursive: w.nonRecursive})
```

Comment:
If you've seen a non-recursive watcher followed by a recursive watcher, would it be better to only rely on the recursive watcher glob?

Maybe instead of a `Set[glob]` you use something like a `map[string]bool` to track directories, but map to whether a recursive glob has ever been encountered.